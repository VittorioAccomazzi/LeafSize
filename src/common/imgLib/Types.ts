
/**
 * definition of a colour. Each channel is considered an 8 bit integer.
 * the values are *not* pre multiplied by alpha. If alpha is not present 
 * it is considered 255, and so the colour is fully opaque.
 */
export interface Colour {
    r : number,
    g : number,
    b : number,
    a?: number
}

/**
 * 2D point. Defined here because DOMPoint requires many more parameters
 * which are not necessary for the imgLib purposes.
 */
export interface Point {
    x: number,
    y: number
}


/**
 * defines a bondung box.
 */
export interface Bbox {
    ulc : Point,
    size: {
        width : number,
        height: number
    }
}

/**
 * generic interface for an image/mask object. Something which can be displayed and/or overlayed on an image.
 */
export default interface IImage<T> {
    readonly width : number,
    readonly height: number,
    set : (x:number, y:number, v : T )=> void,
    get : (x:number, y:number) => T,
    foreachPixel : (func:(x:number, y:number, v:T)=>void)=>void
}

/* unfortunately this is necesary in order to deal with the jest dom, which doesn't support the
 * ImageData object and the browser dom.
 */

export type ImageDataAllocator = ( data : Uint8ClampedArray, width : number, height : number ) => ImageData;

export const imageDataAllocator = ( data : Uint8ClampedArray, width : number, height : number ) : ImageData =>{
    return new ImageData(data, width, height);
}