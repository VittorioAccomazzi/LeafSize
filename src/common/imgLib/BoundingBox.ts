import { Bbox, Point } from "./Types";
import { ImageDataAllocator, imageDataAllocator } from "./Types";

/**
 * simple class to manipulate bounding boxes.
 */
export default class BoundingBox  {

    static allocator = imageDataAllocator;

    private constructor () { }  // singleton

    /**
     * Set the allocator for this class. Use for testing.
     * @param imageDataAllocator functon which instiate the canvas's imagedata
     */
    static SetImageDataAllocator( imageDataAllocator  : ImageDataAllocator  ) : void {
        BoundingBox.allocator=imageDataAllocator
    }

    static FromPoints( pt1 : Point, pt2 : Point ){
        return BoundingBox.FromValues(pt1.x, pt1.y, pt2.x, pt2.y);
    }

    static FromValues( x1 : number, y1 : number, x2 : number, y2 : number ) : Bbox {
        const xmin = Math.min(x1, x2);
        const xmax = Math.max(x1, x2);
        const ymin = Math.min(y1, y2);
        const ymax = Math.max(y1, y2);
        return {
            ulc : {
                x: xmin,
                y: ymin
            },
            size : {
                width : xmax-xmin,
                height: ymax-ymin
            }
        }
    }

    static FromImage(imageData : ImageData) : Bbox {
        return {
            ulc : { x:0, y:0 },
            size: { width:imageData.width, height:imageData.height}
        }
    }
    
    static get Empty() : Bbox {
        return {
            ulc : {x:0, y:0},
            size: {width:0, height:0}
        }
    }

    static IsEmpty(bbox:Bbox) : boolean {
        return bbox.size.width <=0 || bbox.size.height <= 0;
    }

    static Merge( boxes : Bbox[] ) : Bbox {
        let box =  BoundingBox.Empty;
        if( boxes.length > 0 ){
            let xmin = boxes[0].ulc.x;
            let ymin = boxes[0].ulc.y;
            let xmax = xmin;
            let ymax = ymin;
            boxes.forEach(b=>{
                xmin = Math.min(b.ulc.x, xmin);
                ymin = Math.min(b.ulc.y, ymin);
                xmax = Math.max(b.ulc.x+b.size.width, xmax);
                ymax = Math.max(b.ulc.y+b.size.height, ymax);
            })
            box = BoundingBox.FromValues(xmin, ymin, xmax, ymax);
        }
        return box;
    }

    static Dilate(bbox:Bbox, dx : number, dy: number ): Bbox{
        return {
            ulc : { x: bbox.ulc.x-dx, y:bbox.ulc.y-dy},
            size: { width : bbox.size.width+2*dx, height : bbox.size.height+2*dy}
        }
    }

    static Intersect(boxes : Bbox[] ) : Bbox {
        let box =  BoundingBox.Empty;
        if( boxes.length > 0 ){
            let xmin = boxes[0].ulc.x;
            let ymin = boxes[0].ulc.y;
            let xmax = xmin+boxes[0].size.width;
            let ymax = ymin+boxes[0].size.height;
            boxes.forEach(b=>{
                xmin = Math.max(b.ulc.x, xmin);
                ymin = Math.max(b.ulc.y, ymin);
                xmax = Math.min(b.ulc.x+b.size.width, xmax);
                ymax = Math.min(b.ulc.y+b.size.height, ymax);
            })
            box = {
                ulc: {x:xmin, y:ymin},
                size:{width:xmax-xmin, height:ymax-ymin}
            }
        }
        return box;
    }

    static CropImage(bbox : Bbox, imageData : ImageData ) : ImageData | null {
        const imageBox = BoundingBox.FromImage(imageData)
        
        const int = BoundingBox.Intersect([bbox, imageBox]);
        let res : ImageData | null = null;

        if( !BoundingBox.IsEmpty(int) ){
            const width = int.size.width;
            const height= int.size.height;
            const dst = new Uint8ClampedArray(width*height*4);
            const src = imageData.data;
            let dstPt = 0;

            for( let y=int.ulc.y; y<int.ulc.y+height; y++){
                let srcPt= ( y* imageData.width+ bbox.ulc.x ) * 4;
                for( let x=0; x<width; x++ ){
                    dst[dstPt++] = src[srcPt++];
                    dst[dstPt++] = src[srcPt++];
                    dst[dstPt++] = src[srcPt++];
                    dst[dstPt++] = src[srcPt++];
                }
            }

            res = BoundingBox.allocator(dst, width, height);
        }

        return res;
    }

}