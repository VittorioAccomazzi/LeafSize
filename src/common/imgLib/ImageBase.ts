import IImage from "./Types";

interface TypedArray {
    readonly length : number;
    [n: number]: number;
    buffer : ArrayBuffer;
    reduce ( func : (total : number, currentValue : number, index : number ) => number, initialvalue : number) : number;
}

interface TypedArrayConstructor<T extends TypedArray> {
    new (buffer: ArrayBuffer, byteOffset?: number, length?: number): T;
}

export type ImageUint8  = ImageBase<Uint8Array,Uint8ArrayConstructor>
export type ImageUint16 = ImageBase<Uint16Array,Uint16ArrayConstructor>
export type ImageFloat32= ImageBase<Float32Array,Float32ArrayConstructor>

// this is the generic Image type. Using this approach we avoid to expose
// the complexity of the generic used below.

export type Image2D = ImageUint8 | ImageUint16 | ImageFloat32
export type ImagePixels = Uint8Array | Uint16Array | Float32Array
export type ImageType = 'Uint8' | 'Uint16' | 'Float32'


export function isImage( obj : any ) : obj is Image2D {
    return obj.constructor.name === "ImageBase"
}

export class ImageFactory {

    static Uint8( width : number, height : number ) : ImageUint8 {
        return new ImageBase<Uint8Array,Uint8ArrayConstructor>(Uint8Array.BYTES_PER_ELEMENT, Uint8Array, width, height)
    }
    static Uint16( width : number, height : number ) : ImageUint16{
        return (new ImageBase<Uint16Array,Uint16ArrayConstructor>(Uint16Array.BYTES_PER_ELEMENT, Uint16Array, width, height)) 
    }
    static Float32( width : number, height : number ): ImageFloat32{
        return new ImageBase<Float32Array,Float32ArrayConstructor>(Float32Array.BYTES_PER_ELEMENT, Float32Array, width, height)
    }
    static Image(type : ImageType, width : number, height : number ) : Image2D {
        switch( type ){
            case 'Uint8':
                return ImageFactory.Uint8(width,height)
            case 'Uint16':
                return ImageFactory.Uint16(width,height)
            case 'Float32':
                return ImageFactory.Float32(width,height)
        }
    }
}


class ImageBase<T extends TypedArray, C extends TypedArrayConstructor<T>>  implements IImage<number> {

    private buffer : T
    private pixels : T [] 

    constructor( elementSize : number, constructor : C, width : number, height : number ) {
        let nPixels= width * height;
        let bytes  = nPixels * elementSize
        let buffer= new ArrayBuffer(bytes);
        this.pixels= Array.from(Array(height), (e,i)=> new constructor (buffer, width*elementSize*i, width) ) 
        this.buffer= new constructor(buffer, 0, nPixels)
    }


    /**
     * get type of pixel stored.
     */
    get imageType(): ImageType {
        return this.pixels[0].constructor.name.replace('Array','') as ImageType
    }

    /**
     * returns the current list of pixels.
     */
    get imagePixels(): T {
        return this.buffer 
    }

    /**
     *  get the image buffer
     */
    get imageBuffer() : ArrayBuffer {
        return this.buffer.buffer
    }

    /**
     * get image height
     */
    get height() : number {
        return this.pixels.length
    }

    /**
     * return image height
     */
    get width() : number {
        return this.pixels[0].length
    }

    /*
    * overwrite toString in order to provide information when the image is printed
    */
    toString() : string {
       return `image ${this.imageType} ${this.width}x${this.height}`
    }

    /**
     * get a pixel value.
     * @param x row of the pixel queried
     * @param y column of the pixel queried
     */
    get( x: number, y: number ) : number {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates queried: image size ${this.width}x${this.height} pixel ${x},${y}`)
        return this.pixels[y][x]
    }

    /**
     * loop thought all the pixel in the image
     * @param func function to process each pixel in the image
     */
    foreachPixel (func : (x:number, y:number, v: number )=> void) : void  {
        this.pixels.forEach((val, y)=>{ for(let x=0; x<val.length;x++) func(x,y,val[x])})
    }

    /**
     * max pixel value in the image
     */
    maxValue() : number {
        return this.buffer.reduce((max, val, idx)=>Math.max(val,max),this.buffer[0])
    }

    /**
     * min pixel value in the image
     */
    minValue() : number {
        return this.buffer.reduce((min, val, idx)=>Math.min(val,min),this.buffer[0])
    }

    /**
     * get a row of pixels
     * @param y row of pixels to select.
     */
    getRow(y:number) : T {
        if( y<0 || y>this.height )throw new Error (`Invalid y coordinate queried : image height ${this.height} y ${y}`)
        return this.pixels[y]
    }

    /**
     * 
     * @param x row pixel to set
     * @param y column pixel to se
     * @param value  value to set
     */
    set(x:number, y:number, value : number) : void {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates set: image size ${this.width}x${this.height} pixel ${x},${y}`)
        this.pixels[y][x]= value
    }

    /**
     * convert and copy this image in a new image 8bit
     * @param slope multiplication factor applied to every pixel in the conversion
     * @param intercept intercept used for every pixel in the conversion
     */
    convertTo8Uint( slope? : number, intercept? : number) : ImageUint8 {
        return this.convertTo<ImageUint8>('Uint8', slope, intercept);
    }

    /**
     * convert and copy this image in a new image 16bits
     * @param slope multiplication factor applied to every pixel in the conversion
     * @param intercept intercept used for every pixel in the conversion
     */
    convertTo16Uint( slope? : number, intercept? : number) : ImageUint16 {
        return this.convertTo<ImageUint16>('Uint16', slope, intercept);
    }

    /**
     * convert and copy this image in a new image 16bits
     * @param slope multiplication factor applied to every pixel in the conversion
     * @param intercept intercept used for every pixel in the conversion
     */
    convertTo32Float( slope? : number, intercept? : number) : ImageFloat32 {
        return this.convertTo<ImageFloat32>('Float32', slope, intercept);
    }

    private convertTo<ImgType extends Image2D> (type : ImageType, slope? : number, intercept? : number) : ImgType {
        let dstImage = ImageFactory.Image(type, this.width, this.height)
        let dstPixels= dstImage.imagePixels
        if( slope !=null  && intercept != null ){
            dstPixels.forEach((v :number,i:number)=>dstPixels[i]=slope * this.buffer[i]+intercept)
        } else {
            dstPixels.forEach((v :number,i:number)=>dstPixels[i]=this.buffer[i])
        }
        return dstImage as ImgType
    }

    /**
     * check that the value passed are valid image coordinates.
     * @param x row
     * @param y column
     */
    private checkBoundary(x: number, y : number ) : boolean {
        return x>=0 && x < this.width && y>=0 && y<this.height
    }

}