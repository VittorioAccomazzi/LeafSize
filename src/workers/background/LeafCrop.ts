import { dilation, minHeight, minWidth } from "../../app/const";
import BoundingBox from "../../common/imgLib/BoundingBox";
import { Bbox } from "../../common/imgLib/Types";



/**
 * simple class to crop the image based on the bounding box
 */
export class LeafCrop {
    private constructor() {} // singleton

    static Process( areas : number[], bboxs : Bbox[], imgData : ImageData, scale : number ) {
        const box = BoundingBox.Merge(bboxs);
        let leafData  : ImageData = imgData;
        let leafAreas : number [] = [];
        if( !BoundingBox.IsEmpty(box) ){
            let wDilation = dilation;
            let hDilation = dilation;
            if( box.size.width < minWidth ) wDilation = Math.max(wDilation, ( minWidth-box.size.width)/2|0);
            if( box.size.height< minHeight) hDilation = Math.max(hDilation, ( minHeight-box.size.height)/2|0);
            const dBox = BoundingBox.Dilate(box, wDilation, hDilation);
            leafData = BoundingBox.CropImage(dBox, imgData)!;
            leafAreas= areas.map(v=>(v/(scale*scale)|0));
        }
        return {
            imgData : leafData,
            areas : leafAreas
        };
    }
}