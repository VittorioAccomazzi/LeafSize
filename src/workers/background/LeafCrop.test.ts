import { Canvas } from "canvas";
import BoundingBox from "../../common/imgLib/BoundingBox";
import { Bbox } from "../../common/imgLib/Types";
import hash, { nodeCanvasImageDataAllocator } from "../../common/utils/TestUtils";
import LeafCrop from "./LeafCrop";

describe('LeafCrop', ()=>{

    BoundingBox.SetImageDataAllocator(nodeCanvasImageDataAllocator);

    test('shall crop the image',async ()=> {
        const width = 512;
        const height= 512;
        const canvas = new Canvas(width, height);
        const ctx = canvas.getContext('2d');
        const bbx : Bbox = {ulc: {x:50, y:82}, size:{width:339, height:280}}
        ctx.fillStyle="#FFFFFF";
        ctx.clearRect(0,0,width,height );
        ctx.beginPath();
        ctx.strokeStyle="#0000FF";
        ctx.rect(bbx.ulc.x,bbx.ulc.y, bbx.size.width, bbx.size.height);
        ctx.stroke();
        const imgData = ctx.getImageData(0,0,width,height);
        const res = LeafCrop.Process([10], [bbx], imgData, 1);
        const cnv = new Canvas(res.imgData.width, res.imgData.height );
        const cts = cnv.getContext('2d');
        cts.putImageData(res.imgData, 0,0)
        const hsh = await hash(cnv,"cropped 1 box")
        expect(hsh).toMatchSnapshot();
    })

    test('shall merge bounding boxes',async ()=>{
        const width = 512;
        const height= 512;
        const canvas = new Canvas(width, height);
        const ctx = canvas.getContext('2d');
        const bbx1 : Bbox = {ulc: {x:40, y:52}, size:{width:50, height:98}};
        const bbx2 : Bbox = {ulc: {x:382, y:403}, size:{width:8, height:5}};
        const bbxs = [bbx1, bbx2];
        ctx.fillStyle="#FFFFFF";
        ctx.clearRect(0,0,width,height );
        ctx.beginPath();
        ctx.strokeStyle="#FFFF00";
        bbxs.forEach(bbx => ctx.rect(bbx.ulc.x,bbx.ulc.y, bbx.size.width, bbx.size.height))
        ctx.stroke();
        const imgData = ctx.getImageData(0,0,width,height);
        const res = LeafCrop.Process([10], bbxs, imgData, 1);
        const cnv = new Canvas(res.imgData.width, res.imgData.height );
        const cts = cnv.getContext('2d');
        cts.putImageData(res.imgData, 0,0)
        const hsh = await hash(cnv,"cropped 2 box")
        expect(hsh).toMatchSnapshot();
    })

    test('with empty bounding box shall return the entire image', async ()=>{
        const width = 512;
        const height= 512;
        const canvas = new Canvas(width, height);
        const ctx = canvas.getContext('2d');
        const bbx1 : Bbox = {ulc: {x:40, y:52}, size:{width:250, height:298}};
        const bbx2 : Bbox = {ulc: {x:382, y:403}, size:{width:8, height:5}};
        const bbxs = [bbx1, bbx2];
        ctx.fillStyle="#FFFFFF";
        ctx.clearRect(0,0,width,height );
        ctx.beginPath();
        ctx.strokeStyle="#FF0000";
        bbxs.forEach(bbx => ctx.rect(bbx.ulc.x,bbx.ulc.y, bbx.size.width, bbx.size.height))
        ctx.stroke();
        const imgData = ctx.getImageData(0,0,width,height);
        const res = LeafCrop.Process([10], [], imgData, 1);
        const cnv = new Canvas(res.imgData.width, res.imgData.height );
        const cts = cnv.getContext('2d');
        cts.putImageData(res.imgData, 0,0)
        const hsh = await hash(cnv,"cropped NOT crop the image")
        expect(hsh).toMatchSnapshot();
    })
})