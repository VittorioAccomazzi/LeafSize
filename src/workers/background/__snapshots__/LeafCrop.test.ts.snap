// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeafCrop shall crop the image 1`] = `"fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe00000000000000000000000000007ffe00000000000000000000000000007ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe00000000000000000000000000007ffe00000000000000000000000000007fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"`;

exports[`LeafCrop shall merge bounding boxes 1`] = `"fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe00007ffffffffffffffffffffffffffe00007ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe7ffe7ffffffffffffffffffffffffffe00007ffffffffffffffffffffffffffe00007ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff87ffffffffffffffffffffffffffffff87ffffffffffffffffffffffffffffff87fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"`;

exports[`LeafCrop with empty bounding box shall return the entire image 1`] = `"ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff80000000000000007fffffffffffffff80000000000000007fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff9fffffffffffffff7fffffffffffffff80000000000000007ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"`;
